### DISCUSSION TITLE
Trying to create train/val/test DataSets (by date), for multiple TSs with different start dates

### DISCUSSION THREAD

**Discussion by yuvalarbel** [awslabs/gluonts#2674] [Q&A]
Hi community!

I've been banging my head against the wall for many days, and still can't get this to work. I'd really appreciate any recommendations and help.

I have multiple time series, each beginning at different start times, from the year 2000 up until now (freq='B').
I would like to create a train/val/test split, with training data up until 2017, val up until 2019, and test up to now.
I then need to apply Transformations to these datasets.
Then, I'd like to create a torch DataLoader for each split, that returns randomly-positioned slices of (context_len, prediction_len). I'd like all windows, overlapping (distance/step_size=1).

I've tried many, many combinations of gluonts functions, but nothing has been working. Please give me your advice!
Thanks so much.

**Comment by lostella** [comment]
Hi @yuvalarbel, this seems like a fairly generic task that should be easy to do, but maybe is not well covered by tooling/documentation currently. Let's take it step by step and see what may be missing: I'm assuming your data looks like the following, correct? (It's 10 series in my example)

```python
import pandas as pd
import numpy as np

from gluonts.dataset.pandas import PandasDataset

series = []
for k in range(10):
    index = pd.period_range(
        pd.Period("2000-01-03", freq="B") + 5 * k,
        pd.Period("2023-01-31", freq="B")
    )
    series.append(pd.Series(np.random.uniform(size=len(index)), index=index))

dataset = PandasDataset(series)

print(dataset)

for entry in dataset:
    print(entry)
```

Is that accurate? If so, then the first step you can achieve with `split`:

```python
from gluonts.dataset.split import split

test_dataset = dataset
valid_dataset, _ = split(dataset, date=pd.Period("2019-12-31", freq="B"))
train_dataset, _ = split(dataset, date=pd.Period("2017-12-31", freq="B"))

for a, b, c in zip(train_dataset, valid_dataset, test_dataset):
    print(f"{len(a['target'])} / {len(b['target'])} / {len(c['target'])}")
```

**Reply by yuvalarbel** [reply]
Hi @lostella , thanks for your answer!

Looks great (very similar to my use case). I have two unsolved problems:

1) Your generated data looks similar to mine, except the fact that some of my time series begin only after the val/test sets cutoffs, so don't exist in the training/val sets. This creates a problem when using the train/val sets, as illustrated in the following recreation code:

```python
import pandas as pd
import numpy as np

from gluonts.dataset.pandas import PandasDataset
from gluonts.dataset.split import split


FREQUENCY = 'B'

# START_TIME_DELAY_DIFF = 5
START_TIME_DELAY_DIFF = int(260 * 2.5)  # two and a half years


series = []
for k in range(10):
    index = pd.period_range(
        pd.Period("2000-01-03", freq=FREQUENCY) + (START_TIME_DELAY_DIFF + k) * k,
        pd.Period("2023-01-31", freq=FREQUENCY)
    )
    series.append(pd.Series(np.random.uniform(size=len(index)), index=index))

dataset = PandasDataset(series)

print("Start times:")
for entry in dataset:
    print(entry['start'])

test_dataset = dataset
valid_dataset, _ = split(dataset, date=pd.Period("2019-12-31", freq=FREQUENCY))
train_dataset, _ = split(dataset, date=pd.Period("2017-12-31", freq=FREQUENCY))

for entry in train_dataset:
    print(entry)
```

This results in this AssertionError:
```
Start times:
2000-01-03
2002-07-02
2004-12-31
2007-07-06
2010-01-12
2012-07-23
2015-02-03
2017-08-18
2020-03-06
2022-09-27
2000-01-03
2002-07-02
2004-12-31
2007-07-06
2010-01-12
2012-07-23
2015-02-03
2017-08-18
2020-03-06

Traceback (most recent call last):
  Input line 47, in <module>
    for entry in train_dataset:
  File "gluonts\dataset\split.py", line 258, in generate_training_entries
    yield from map(self.training_entry, dataset)
  File "gluonts\dataset\split.py", line 349, in training_entry
    return slice_data_entry(entry, slice(None, length))
  File "gluonts\dataset\split.py", line 161, in slice_data_entry
    slice_ = to_positive_slice(
  File "gluonts\dataset\split.py", line 118, in to_positive_slice
    assert stop >= 0
AssertionError
```

2) From the validation/test datasets, I'd like to only sample examples in which the context and prediction targets are contained in the date slice (so val dataset should give examples with context/prediction targets inside years 2018-2019, and test 2020-today)
I could not find a sampler that is able to discriminate between dates. I could only use the `TestTemplate` returned by the `split()` function, but could not find how to (a) get all instances in desired years (I have to give `template.generate_instances()` a specific number of windows to generate) and (b) compute transforms on these templates, then aggregate these all into a dataset, sampling randomly from them.
Hopefully I could find and use a variation of `ExpectedNumInstanceSampler`, only sampling from the right slice of dates.

Eagerly awaiting your reply for both of these remaining problems, and thanks again for the help!

**Reply by yuvalarbel** [reply]
Hey,

After many forceful wall-to-head iteractions, I finally found a possible solution. I have created new gluonts sampler and splitter objects to solve these two problems.

In order to do that, I had to pass the `sampler.__call__` method and the `sampler._get_bounds` method the start Period as well as the timeseries itself. (The signature of these two methods now do not match the signature of the respective methods in `InstanceSampler`.)

The only change I made in the InstanceSplitter is to also pass `data[self.start_field]` to the sampler. The sampler uses this start time to change the bounds returned from `sampler._get_bounds`.

One last change I made to the ExpectedNumInstanceSampler is to have the expected number of instances from each dataset be proportional to the log of the window size (window size usually being the length of the timeseries, but altered according to the start/end dates given and min_past/future given). This makes sure a more even representation is given between the timeseries in the dataset.

This is the code to the two classes I created:

```python
from typing import Tuple, Iterator
import numpy as np
import pandas as pd

from gluonts.transform import InstanceSampler, InstanceSplitter
from gluonts.dataset.common import DataEntry
from gluonts.dataset.split import periods_between


class DateSplittedLogExpectedNumInstanceSampler(InstanceSampler):
    num_instances: float
    start_date: pd.Period = None
    end_date: pd.Period = None
    total_length: int = 0
    n: int = 0

    def _get_bounds(self, ts: np.ndarray, start: pd.Period) -> Tuple[int, int]:

        begin = self.min_past
        window_end = ts.shape[self.axis]

        if self.start_date is not None and start < self.start_date:
            begin += periods_between(start, self.start_date) - 1

        if self.end_date is not None:
            if self.end_date < start:
                return begin, -1
            periods = periods_between(start, self.end_date)
            if periods < window_end:
                window_end = periods

        end = window_end - self.min_future

        return begin, end

    def __call__(self, ts: np.ndarray, start: pd.Period) -> np.ndarray:
        # Should return indices of the beginnings of the future windows

        a, b = self._get_bounds(ts, start)
        window_size = b - a + 1

        if window_size <= 0:
            return np.array([], dtype=int)

        self.n += 1
        self.total_length += window_size
        avg_length = self.total_length / self.n

        if avg_length <= 0:
            return np.array([], dtype=int)

        log_avg_length = np.log(avg_length)
        log_window_size = np.log(window_size)

        expected_num_instances = self.num_instances * log_window_size / log_avg_length

        p = expected_num_instances / avg_length
        (indices,) = np.where(np.random.random_sample(window_size) < p)
        return indices + a


class DateInstanceSplitter(InstanceSplitter):
    def flatmap_transform(
        self, data: DataEntry, is_train: bool
    ) -> Iterator[DataEntry]:
        pl = self.future_length
        lt = self.lead_time
        slice_cols = self.ts_fields + [self.target_field]
        target = data[self.target_field]
        start = data[self.start_field]

        sampled_indices = self.instance_sampler(target, start)

        for i in sampled_indices:
            pad_length = max(self.past_length - i, 0)
            d = data.copy()
            for ts_field in slice_cols:
                if i > self.past_length:
                    # truncate to past_length
                    past_piece = d[ts_field][..., i - self.past_length : i]
                elif i < self.past_length:
                    pad_block = (
                        np.ones(
                            d[ts_field].shape[:-1] + (pad_length,),
                            dtype=d[ts_field].dtype,
                        )
                        * self.dummy_value
                    )
                    past_piece = np.concatenate(
                        [pad_block, d[ts_field][..., :i]], axis=-1
                    )
                else:
                    past_piece = d[ts_field][..., :i]
                d[self._past(ts_field)] = past_piece
                d[self._future(ts_field)] = d[ts_field][
                    ..., i + lt : i + lt + pl
                ]
                del d[ts_field]
            pad_indicator = np.zeros(self.past_length, dtype=target.dtype)
            if pad_length > 0:
                pad_indicator[:pad_length] = 1

            if self.output_NTC:
                for ts_field in slice_cols:
                    d[self._past(ts_field)] = d[
                        self._past(ts_field)
                    ].transpose()
                    d[self._future(ts_field)] = d[
                        self._future(ts_field)
                    ].transpose()

            d[self._past(self.is_pad_field)] = pad_indicator
            d[self.forecast_start_field] = d[self.start_field] + i + lt
            yield d
```

How to use this sampler/splitter pair for my data:
```python
from gluonts.dataset.field_names import FieldName
from gluonts.torch.util import IterableDataset

FREQUENCY = 'B'
CONTEXT_LENGTH = 3
PREDICTION_LENGTH = 2

# Example for validation set times (you don't need to give both start and end)
sampler = DateSplittedLogExpectedNumInstanceSampler(
                num_instances=1000,
                min_past=CONTEXT_LENGTH,
                min_future=PREDICTION_LENGTH,
                start_date=pd.Period("2018-01-01", freq=FREQUENCY),
                end_date=pd.Period("2019-12-31", freq=FREQUENCY)
            )

splitter = DateInstanceSplitter(
                target_field=FieldName.TARGET,
                is_pad_field=FieldName.IS_PAD,
                start_field=FieldName.START,
                forecast_start_field=FieldName.FORECAST_START,
                instance_sampler=sampler,
                past_length=CONTEXT_LENGTH,
                future_length=PREDICTION_LENGTH,
                time_series_fields=[FieldName.FEAT_TIME, FieldName.OBSERVED_VALUES],
            )

dataset = IterableDataset(splitter.apply(dataset, is_train=True))
```

Tell me what you think.
Thanks!

**Reply by lostella** [reply]
Hi @yuvalarbel, I think your solution makes perfect sense. In fact, achieving what you need appears otherwise hard with the current tooling:
- `split` works by date, but 1) only operates on the "input schema" and otherwise fails if you massage your data to add or rename fields; it's mostly designed to make sure one can split the dataset so that train/valid/test data are "consistent" (ie there's no overlap between training data and validation labels). The idea however is that the instance splitter takes care of sampling and slicing windows (for whatever purpose).
- The instance splitter logic is currently by index only. So to achieve what you need, one would make sure the two (`split` and instance splitter) are used together the right way, which is probably hard to do and error prone.

So, some more direct mechanism to directly express "I want windows between this and that date from each series" (either randomly sampled, or just all of them) may be actually good to have.

**Reply by lostella** [reply]
Let me open an issue about this, where I can jot down my thoughts on what a good general tool there would be

**Reply by yuvalarbel** [reply]
Great, thanks @lostella !
I'd love to get updates on this issue :)
Waiting to see which general tool will be added!
Cheers.

**Reply by yuvalarbel** [reply]
Hello @lostella , has an issue been opened?

**Reply by lostella** [reply]
@yuvalarbel I gave it some thought, and made a proposal [here](https://github.com/awslabs/gluonts/pull/2758) (draft, feedback is appreciated)
